compiler/resolution/functionResolution.cpp

```diff
@@ -1525,16 +1538,21 @@ userCall(CallExpr* call) {
   // If the called function is compiler-generated or is in one of the internal
   // modules, back up the stack until a call is encountered whose target
   // function is neither.
-  // TODO: This function should be rewritten so each test appears only once.
-  if (call->getFunction()->hasFlag(FLAG_COMPILER_GENERATED) ||
-      call->getModule()->modTag == MOD_INTERNAL) {
-    for (int i = callStack.n-1; i >= 0; i--) {
-      if (!callStack.v[i]->getFunction()->hasFlag(FLAG_COMPILER_GENERATED) &&
-          callStack.v[i]->getModule()->modTag != MOD_INTERNAL)
-        return callStack.v[i];
-    }
+
+  CallExpr* cur = call;
+  int i = callStack.n-1;
+  while (cur != NULL && i >= 0 && shouldSkip(cur)) {
+    cur = callStack.v[i];
+    i -= 1;
+  }
```

Comment:
I've been doing the transformation from the old code to the refactored code in my head and I'm worried there's an off-by-one error in the call stack accesses. I don't think there is but it's harder to follow than the old version. In some ways I'd prefer the structure

``` c++
if (shouldSkip(call)) {
  for (int i = callStack.n-1; i >= 0; i--) {
     CallExpr* cur = callStack.v[i];
     if (!shouldSkip(cur))
       return cur;
  }
}
return call;
```